Regression: Generating namespace packages is broken

cookieplone 0.9.0 and 0.9.1 changed how Python namespace packages are generated to use pkgutil, but they are still listed as setuptools namespace packages in setup.py (https://github.com/plone/cookieplone-templates/blob/main/backend_addon/%7B%7B%20cookiecutter.__folder_name%20%7D%7D/setup.py#L49).

As a result, trying to generate a namespace package (such as collective.something) and then run `make test` fails like this:

```
test: packaging backend failed (code=error: Namespace package problem: collective is a namespace package, but its
__init__.py does not call declare_namespace()! Please fix it.
(See the setuptools manual under "Namespace Packages" for details.)
"), with SystemExit: error: Namespace package problem: collective is a namespace package, but its
__init__.py does not call declare_namespace()! Please fix it.
(See the setuptools manual under "Namespace Packages" for details.)
"
/Users/davisagli/Plone/collective-something/backend/.tox/.pkg/lib/python3.12/site-packages/setuptools/dist.py:701: SetuptoolsDeprecationWarning: The namespace_packages parameter is deprecated.
!!

        ********************************************************************************
        Please replace its usage with implicit namespaces (PEP 420).

        See https://setuptools.pypa.io/en/latest/references/keywords.html#keyword-namespace-packages for details.
        ********************************************************************************

!!
  ep.load()(self, ep.name, value)
Traceback (most recent call last):
  File "/Users/davisagli/Plone/collective-something/backend/.tox/.pkg/lib/python3.12/site-packages/setuptools/_distutils/core.py", line 202, in run_commands
    dist.run_commands()
  File "/Users/davisagli/Plone/collective-something/backend/.tox/.pkg/lib/python3.12/site-packages/setuptools/_distutils/dist.py", line 983, in run_commands
    self.run_command(cmd)
  File "/Users/davisagli/Plone/collective-something/backend/.tox/.pkg/lib/python3.12/site-packages/setuptools/dist.py", line 999, in run_command
    super().run_command(command)
  File "/Users/davisagli/Plone/collective-something/backend/.tox/.pkg/lib/python3.12/site-packages/setuptools/_distutils/dist.py", line 1002, in run_command
    cmd_obj.run()
  File "/Users/davisagli/Plone/collective-something/backend/.tox/.pkg/lib/python3.12/site-packages/setuptools/command/egg_info.py", line 312, in run
    self.find_sources()
  File "/Users/davisagli/Plone/collective-something/backend/.tox/.pkg/lib/python3.12/site-packages/setuptools/command/egg_info.py", line 320, in find_sources
    mm.run()
  File "/Users/davisagli/Plone/collective-something/backend/.tox/.pkg/lib/python3.12/site-packages/setuptools/command/egg_info.py", line 543, in run
    self.add_defaults()
  File "/Users/davisagli/Plone/collective-something/backend/.tox/.pkg/lib/python3.12/site-packages/setuptools/command/egg_info.py", line 581, in add_defaults
    sdist.add_defaults(self)
  File "/Users/davisagli/Plone/collective-something/backend/.tox/.pkg/lib/python3.12/site-packages/setuptools/command/sdist.py", line 109, in add_defaults
    super().add_defaults()
  File "/Users/davisagli/Plone/collective-something/backend/.tox/.pkg/lib/python3.12/site-packages/setuptools/_distutils/command/sdist.py", line 237, in add_defaults
    self._add_defaults_python()
  File "/Users/davisagli/Plone/collective-something/backend/.tox/.pkg/lib/python3.12/site-packages/setuptools/command/sdist.py", line 121, in _add_defaults_python
    self.filelist.extend(build_py.get_source_files())
                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/davisagli/Plone/collective-something/backend/.tox/.pkg/lib/python3.12/site-packages/setuptools/_distutils/command/build_py.py", line 303, in get_source_files
    return [module[-1] for module in self.find_all_modules()]
                                     ^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/davisagli/Plone/collective-something/backend/.tox/.pkg/lib/python3.12/site-packages/setuptools/_distutils/command/build_py.py", line 298, in find_all_modules
    m = self.find_package_modules(package, package_dir)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/davisagli/Plone/collective-something/backend/.tox/.pkg/lib/python3.12/site-packages/setuptools/_distutils/command/build_py.py", line 220, in find_package_modules
    self.check_package(package, package_dir)
  File "/Users/davisagli/Plone/collective-something/backend/.tox/.pkg/lib/python3.12/site-packages/setuptools/command/build_py.py", line 261, in check_package
    raise distutils.errors.DistutilsError(
distutils.errors.DistutilsError: Namespace package problem: collective is a namespace package, but its
__init__.py does not call declare_namespace()! Please fix it.
(See the setuptools manual under "Namespace Packages" for details.)
"

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/davisagli/Plone/collective-something/backend/.venv/lib/python3.12/site-packages/pyproject_api/_backend.py", line 93, in run
    outcome = backend_proxy(parsed_message["cmd"], **parsed_message["kwargs"])
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/davisagli/Plone/collective-something/backend/.venv/lib/python3.12/site-packages/pyproject_api/_backend.py", line 34, in __call__
    return getattr(on_object, name)(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/davisagli/Plone/collective-something/backend/.tox/.pkg/lib/python3.12/site-packages/setuptools/build_meta.py", line 337, in get_requires_for_build_sdist
    return self._get_build_requires(config_settings, requirements=[])
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/davisagli/Plone/collective-something/backend/.tox/.pkg/lib/python3.12/site-packages/setuptools/build_meta.py", line 304, in _get_build_requires
    self.run_setup()
  File "/Users/davisagli/Plone/collective-something/backend/.tox/.pkg/lib/python3.12/site-packages/setuptools/build_meta.py", line 522, in run_setup
    super().run_setup(setup_script=setup_script)
  File "/Users/davisagli/Plone/collective-something/backend/.tox/.pkg/lib/python3.12/site-packages/setuptools/build_meta.py", line 320, in run_setup
    exec(code, locals())
  File "<string>", line 15, in <module>
  File "/Users/davisagli/Plone/collective-something/backend/.tox/.pkg/lib/python3.12/site-packages/setuptools/__init__.py", line 117, in setup
    return distutils.core.setup(**attrs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/davisagli/Plone/collective-something/backend/.tox/.pkg/lib/python3.12/site-packages/setuptools/_distutils/core.py", line 186, in setup
    return run_commands(dist)
           ^^^^^^^^^^^^^^^^^^
  File "/Users/davisagli/Plone/collective-something/backend/.tox/.pkg/lib/python3.12/site-packages/setuptools/_distutils/core.py", line 216, in run_commands
    raise SystemExit("error: " + str(msg))
SystemExit: error: Namespace package problem: collective is a namespace package, but its
__init__.py does not call declare_namespace()! Please fix it.
(See the setuptools manual under "Namespace Packages" for details.)
"
Backend: run command get_requires_for_build_sdist with args {'config_settings': None}
running egg_info
creating src/collective.something.egg-info
writing src/collective.something.egg-info/PKG-INFO
writing dependency_links to src/collective.something.egg-info/dependency_links.txt
writing entry points to src/collective.something.egg-info/entry_points.txt
writing namespace_packages to src/collective.something.egg-info/namespace_packages.txt
writing requirements to src/collective.something.egg-info/requires.txt
writing top-level names to src/collective.something.egg-info/top_level.txt
writing manifest file 'src/collective.something.egg-info/SOURCES.txt'
Backend: Wrote response {'code': 'error: Namespace package problem: collective is a namespace package, but its\n__init__.py does not call declare_namespace()! Please fix it.\n(See the setuptools manual under "Namespace Packages" for details.)\n"', 'exc_type': 'SystemExit', 'exc_msg': 'error: Namespace package problem: collective is a namespace package, but its\n__init__.py does not call declare_namespace()! Please fix it.\n(See the setuptools manual under "Namespace Packages" for details.)\n"'} to /var/folders/1n/mxt0jv4d5yqdgx3kfq6lnbxh0000gn/T/pep517_get_requires_for_build_sdist-4ejecypu.json
  test: FAIL code error: Namespace package problem: collective is a namespace package, but its
__init__.py does not call declare_namespace()! Please fix it.
(See the setuptools manual under "Namespace Packages" for details.)
" (35.69 seconds)
  evaluation failed :( (35.78 seconds)
error: Namespace package problem: collective is a namespace package, but its
__init__.py does not call declare_namespace()! Please fix it.
(See the setuptools manual under "Namespace Packages" for details.)
"
make: *** [test] Error 1
```

--- a/cookieplone/utils/plone.py
+++ b/cookieplone/utils/plone.py
@@ -58,15 +58,24 @@ def format_python_codebase(path: Path):
         func(path)
 
 
-def create_namespace_packages(path: Path, package_name: str):
+def create_namespace_packages(
+    path: Path, package_name: str, style: str = "pkg_resources"
+):
     """Create namespace packages to hold an existing package."""
     current = path.parent
     for namespace in package_name.split(".")[:-1]:
         current = current / namespace
         current.mkdir()
-        # Use pkgutils, as new packages are not setuputils based
-        (current / "__init__.py").write_text(
-            "from pkgutil import extend_path\n\n"
-            "__path__ = extend_path(__path__, __name__)\n"
-        )
+        if style == "pkg_resources":
+            (current / "__init__.py").write_text(
+                '__import__("pkg_resources").declare_namespace(__name__)\n'
+            )
+        elif style == "pkgutil":
+            (current / "__init__.py").write_text(
+                "from pkgutil import extend_path\n\n"
+                "__path__ = extend_path(__path__, __name__)\n"
+            )
+        else:
+            # PEP 420 implicit namespace; no file needed.
+            pass
     path.rename(current / package_name.split(".")[-1])

--- a/news/80.bugfix
+++ b/news/80.bugfix
@@ -1,0 +1,1 @@
+Fix regression in generating namespace packages. @davisagli

